### PR TITLE
Add AsArray trait for more ergonomic downcasting

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

We currently provide downcasting utility functions, however, they have a couple of issues:

* Lack of method chaining, makes for more nested and potentially less legible code
* Methods are named inconsistently and are unnecessarily verbose - e.g `as_largestring_array` but `as_large_list_array`
* Not all methods can easily be used in generic context, e.g. no `as_byte_array`
* When called on `&ArrayRef` result in two dynamic dispatches, as `ArrayRef` is itself coerced to `dyn Array`
* Lack of autoderef means you often need to add `&` as necessary
* Each method has to be imported separately

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/builder/generic_byte_run_builder.rs [hunk 1]
```diff
@@ -368,7 +368,7 @@ pub type LargeStringRunBuilder<K> = GenericByteRunBuilder<K, LargeUtf8Type>;
 ///
 /// // Values are polymorphic and so require a downcast.
 /// let av = array.values();
-/// let ava: &BinaryArray = as_generic_binary_array::<i32>(av.as_ref());
+/// let ava: &BinaryArray = av.as_binary();
```
Comment (alamb) [inline arrow-array/src/builder/generic_byte_run_builder.rs:371]: this is much nicer -- I hit this with @appletreeisyellow when trying to downcast to a `BinaryArray` (having to know to put i32 was non obvious)

File: arrow-ord/src/comparison.rs [hunk 2]
```diff
@@ -893,14 +875,8 @@ pub fn lt_eq_dyn_binary_scalar(
...
     match left.data_type() {
-        DataType::Binary => {
-            let left = as_generic_binary_array::<i32>(left);
-            lt_eq_binary_scalar(left, right)
-        }
-        DataType::LargeBinary => {
-            let left = as_generic_binary_array::<i64>(left);
-            lt_eq_binary_scalar(left, right)
-        }
+        DataType::Binary => lt_eq_binary_scalar(left.as_binary::<i32>(), right),
```
Comment (alamb) [inline arrow-ord/src/comparison.rs:878]: it sure looks nicer

File: arrow-array/src/cast.rs [hunk 3]
```diff
@@ -709,6 +709,165 @@ where
...
+/// # use std::sync::Arc;
+/// # use arrow_array::{ArrayRef, Int32Array};
+/// # use arrow_array::cast::AsArray;
+/// # use arrow_array::types::Int32Type;
+/// let col = Arc::new(Int32Array::from(vec![1, 2, 3])) as ArrayRef;
+/// assert_eq!(col.as_primitive::<Int32Type>().values(), &[1, 2, 3]);
+/// ```
+pub trait AsArray: private::Sealed {
+    /// Downcast this to a [`BooleanArray`] returning `None` if not possible
+    fn as_boolean_opt(&self) -> Option<&BooleanArray>;
```
Comment (alamb) [inline arrow-array/src/cast.rs:728]: I like how you have provided both `_opt` and non `_opt` variants so users can choose

File: arrow-array/src/cast.rs [hunk 4]
```diff
@@ -709,6 +709,165 @@ where
...
+    /// Downcast this to a [`DictionaryArray`] panicking if not possible
+    fn as_dictionary<K: ArrowDictionaryKeyType>(&self) -> &DictionaryArray<K> {
+        self.as_dictionary_opt().expect("dictionary array")
+    }
+}
+
+impl private::Sealed for dyn Array + '_ {}
+impl AsArray for dyn Array + '_ {
+    fn as_boolean_opt(&self) -> Option<&BooleanArray> {
+        self.as_any().downcast_ref()
```
Comment (alamb) [inline arrow-array/src/cast.rs:808]: this is very clever 👍

**Review by alamb — APPROVED** [review]
LGTM

Some other suggestions:
1. Mark `as_XXX_array` functions deprecated (or maybe make a note to do so in a later release)
2. Consider updating the example in the main README: https://docs.rs/arrow/35.0.0/arrow/#type-erasure--trait-objects

File: arrow-array/src/builder/generic_byte_run_builder.rs (hunk 1 above)
Comment (appletreeisyellow) [reply arrow-array/src/builder/generic_byte_run_builder.rs:371]: Awesome!

Comment (tustvold) [reply arrow-array/src/builder/generic_byte_run_builder.rs:371]: Err... The i32 is just being inferred because of the `: &BinaryArray` type constraint...

**Comment by tustvold** [comment]
> Mark as_XXX_array functions deprecated (or maybe make a note to do so in a later release)

I will make a note to do so in a future PR

File: arrow-array/src/builder/generic_byte_run_builder.rs (hunk 1 above)
Comment (alamb) [reply arrow-array/src/builder/generic_byte_run_builder.rs:371]: I am thinking as a user here 🤔

Comment (tustvold) [reply arrow-array/src/builder/generic_byte_run_builder.rs:371]: There is also `as_bytes::<Utf8Type>()` if that is more obvious for users?

Comment (alamb) [reply arrow-array/src/builder/generic_byte_run_builder.rs:371]: I don't necessarily think the code needs to change. I think what helps the most is having some examples around that you can start from. So as always more docs / examples I think the better

**Pull request merged**